added a note on PVC usage (#224)
stackabletech/documentation: modules/concepts/pages/pvc.adoc

= `PersistentVolumeClaim` usage

Several of the tools on the Stackable platform can use external resources that the cluster administrator makes available via a `PersistentVolume`. Airflow users can access DAG jobs this way, and Spark users can do the same for data or other job dependencies, to name just two examples.

> A `PersistentVolume` will usually be provisioned by the cluster administrator, who will take into account the type of storage that is required. This will include, for example, an appropriate sizing, and relevant access modes (which  in turn are dependent on the `StorageClass` chosen to back the `PersistentVolume`).

Author: Changed as per suggestion

> A `PersistentVolume` will usually be provisioned by the cluster administrator, who will take into account the type of storage that is required. This will include, for example, an appropriate sizing, and relevant access modes (which  in turn are dependent on the `StorageClass` chosen to back the `PersistentVolume`).

Author: Changed as per suggestion

== Access modes and the `StorageClass`

Not all access modes are compatible with all storage classes. This https://docs.openshift.com/container-platform/4.8/storage/understanding-persistent-storage.html#pv-access-modes_understanding-persistent-storage[link] provides a helpful compatability matrix (in this case for OpenShift). Other managed Kubernetes implementations will be similar, albeit with different default storage class names. The important point is that the default `StorageClass` only supports `ReadWriteOnce`, which limits access to the `PersistentVolumeClaim` to a single node.

== Node selection

The Kubernetes https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/[documentation] states the following with regard to assigning pods to specific nodes:
____
Generally such constraints are unnecessary, as the scheduler will automatically do a reasonable placement (for example, spreading your Pods across nodes so as not place Pods on a node with insufficient free resources).
____

> This suggests that resources are automatically considered when pods are assigned to nodes, but it is not clear if the same is true for implicit dependencies, such as `PersistentVolumeClaim` usage. The https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/[scheduler] will take various factors into account, such as

Reviewer: Yes. The PVC is resolved to its corresponding PV, and the scheduling constraints of the PV are taken into account when scheduling the node.

Author: Changed as per suggestion

____
...individual and collective resource requirements, hardware / software / policy constraints, affinity and anti-affinity specifications, data locality, inter-workload interference....
____

but implementations may vary in the way soft dependencies (e.g. optimal resource usage) and hard dependencies (e.g. access modes, that may prevent the job from running) are handled and prioritised.

== Test considerations

> For `PersistentVolumeClaim`-relevant tests in the Stackable operator repositories the backing `PersistentVolume` is usually omitted as this is an implementation/provisioning choice to made by the cluster administrator and mocking e.g. an NFS volume for tests is non-trivial.

Author: Changed as per suggestion

> The only viable access mode is thus `ReadWriteOnce` (see above) which means that all test steps dependent on a `PersistentVolumeClaim` should be run on the same node, and this assignment should be made explicitly with a declaration of either a node selector or pod-affinity.

Author: Changed as per suggestion

> The only viable access mode is thus `ReadWriteOnce` (see above) which means that all test steps dependent on a `PersistentVolumeClaim` should be run on the same node, and this assignment should be made explicitly with a declaration of either a node selector or pod-affinity.

Author: Marking this as resolved as we have an upcoming meeting to discuss this (with a corresponding fine-tuning of these docs).

> = `PersistentVolumeClaim` usage

Author: Implemented as suggested.

Several of the tools on the Stackable platform can use external resources that the cluster administrator makes available via a `PersistentVolume`. Airflow users can access DAG jobs this way, and Spark users can do the same for data or other job dependencies, to name just two examples.

A `PersistentVolume` will usually be provisioned by the Kubernetes Container Storage Interface (CSI) on behalf of the cluster administrator, who will take into account the type of storage that is required. This will include, for example, an appropriate sizing, and relevant access modes (which  in turn are dependent on the `StorageClass` chosen to back the `PersistentVolume`).

> == Access modes and the `StorageClass`

Author: Added links and an example.

spec:
  storageClassName: manual #<4>
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi #<5>
----
<1> The name of the storage class, which will be used by the PersistentVolumeClaim
<2> The capacity of the PersistentVolume
<3> a list of https://kubernetes.io/docs/concepts/storage/persistent-volumes/?force_isolation=true#access-modes[access modes]
<4> The storageClassName which is used to match a PersistentVolume to a claim
<5> The specific quantity of the resource that is being claimed

== Access modes and the StorageClass

Not all https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes[access modes] are compatible with all storage classes. This https://docs.openshift.com/container-platform/4.8/storage/understanding-persistent-storage.html#pv-access-modes_understanding-persistent-storage[link] provides a helpful compatability matrix (in this case for OpenShift). Other managed Kubernetes implementations will be similar, albeit with different default storage class names. The important point is that the default StorageClass only supports `ReadWriteOnce`, which limits access to the PersistentVolumeClaim to a single node.

> == Node selection

Reviewer 2: This section feels like it is missing a conclusion. Can we find out what to suggest to the user? Should they use labels and selectors to enforce scheduling of pods, or do they not need to do that?

Author: I have added a short list of considerations in this section.

Author: ...and some more links to online documentation and operator examples.

The Kubernetes https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/[documentation] states the following with regard to assigning pods to specific nodes:
____
Generally such constraints are unnecessary, as the scheduler will automatically do a reasonable placement (for example, spreading your Pods across nodes so as not place Pods on a node with insufficient free resources).
____
This suggests that resources are automatically considered when pods are assigned to nodes, but it is not clear if the same is true for implicit dependencies, such as PersistentVolumeClaim usage by multiple pods. The https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/[scheduler] will take various factors into account, such as

____
...individual and collective resource requirements, hardware / software / policy constraints, affinity and anti-affinity specifications, data locality, inter-workload interference....
____

but implementations may vary in the way soft dependencies (e.g. optimal resource usage) and hard dependencies (e.g. access modes, that may prevent the job from running) are handled and prioritised.

> == Test considerations

Author: Yes, I agree. I just wanted to note it somewhere rather than forget it. Shall we leave it here until we find a more obvious place? (maybe under the section for contributors?)

> == Test considerations

Reviewer 2: Yes it's probably fine to leave it here for now

[source,yaml]
----
---
apiVersion: v1
  kind: PersistentVolume
metadata:
  name: task-pv-volume
  labels:
    type: local
spec:
  storageClassName: manual #<1>
  capacity:
    storage: 10Gi #<2>
  accessModes:
    - ReadWriteOnce #<3>
  hostPath:
    path: "/mnt/data"
----

[source,yaml]
----
---
apiVersion: v1
  kind: PersistentVolumeClaim
metadata:
  name: task-pv-claim
spec:
  storageClassName: manual #<4>
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 3Gi #<5>
----
<1> The name of the storage class, which will be used by the PersistentVolumeClaim
<2> The capacity of the PersistentVolume
<3> a list of https://kubernetes.io/docs/concepts/storage/persistent-volumes/?force_isolation=true#access-modes[access modes]
<4> The storageClassName which is used to match a PersistentVolume to a claim
<5> The specific quantity of the resource that is being claimed

== Access modes and the StorageClass

> Not all https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes[access modes] are compatible with all storage classes. This https://docs.openshift.com/container-platform/4.8/storage/understanding-persistent-storage.html#pv-access-modes_understanding-persistent-storage[link] provides a helpful compatability matrix (in this case for OpenShift). Other managed Kubernetes implementations will be similar, albeit with different default storage class names. The important point is that the default StorageClass only supports `ReadWriteOnce`, which limits access to the PersistentVolumeClaim to a single node. A strategy governing PersistentVolumeClaim resources will thus address the following:

Reviewer 2: ```suggestion
Not all storage classes support all https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes[access modes]. The supported access modes also depend on the Kubernetes implementation, see for example the compatiblity table https://docs.openshift.com/container-platform/4.8/storage/understanding-persistent-storage.html#pv-access-modes_understanding-persistent-storage[Supported access modes for PVs] in the OpenShift documentation. Other managed Kubernetes implementations will be similar, albeit with different default storage class names. The important point is that the default StorageClass only supports `ReadWriteOnce`, which limits access to the PersistentVolumeClaim to a single node. A strategy governing PersistentVolumeClaim resources will thus address the following:
```

Reviewer 2: I tried to reformulate to get a better link text: https://developers.google.com/style/link-text

- what access mode is appropriate? Will the resources be accessed by multiple pods and/or modes?
- does the Kubernetes cluster have a default implementation for these access modes?
- if access modes are restricted (e.g. `ReadWriteOnce`), does the cluster prioritise available resources over implicit application dependencies (in other words, is the PersistentVolumeClaim treated as a soft- or hard-dependency)?

If a PersistentVolumeClaim should be mounted on a single node for the application and its components that use it, this can be specified explicitly (see the next section).

== Node selection

The Kubernetes https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/[documentation] states the following with regard to assigning pods to specific nodes:
____

> Generally such constraints are unnecessary, as the scheduler will automatically do a reasonable placement (for example, spreading your Pods across nodes so as not place Pods on a node with insufficient free resources).

Reviewer 2: ```suggestion
the scheduler will automatically do a reasonable placement (for example, spreading your Pods across nodes so as not place Pods on a node with insufficient free resources).
```

Reviewer 2: I removed the "such constraints" bit, because it's referring to something that is not inside the quote, I found that confusing.

____

but implementations may vary in the way soft dependencies (e.g. optimal resource usage) and hard dependencies (e.g. access modes, that may prevent the job from running) are handled and prioritised.

== Test considerations

For PersistentVolumeClaim-relevant tests in the Stackable operator repositories the backing PersistentVolume is omitted as this is an implementation decision to be made by the cluster administrator and mocking e.g. an NFS volume for tests is non-trivial.

If the only viable access mode is `ReadWriteOnce` (see above) - meaning that all test steps dependent on a PersistentVolumeClaim should be run on the same node - this assignment should be made explicitly with a declaration of either a node selector or pod-affinity.

Managed Kubernetes clusters will normally have a default storage implementation for access modes other than `ReadWriteOnce` so e.g. `ReadWriteMany` can be declared for tests running against such clusters in the knowledge that the appropriate storage will be used.

== Operator usage

=== Airflow

> The Airflow operator can use a PersistentVolumeClaim for externally-loaded airflow jobs, known as Direct Acyclic Graphs (DAGs), as shown in this https://github.com/stackabletech/airflow-operator/blob/main/examples/simple-airflow-cluster-dags-pvc.yaml[example]. Airflow expects that all components (webserver, scheduler and workers) have access to the DAG folder, so the PersistentVolumeClaim either has to be accessible with the `ReadWriteMany` access mode or node selection should be declared to ensure all components run on the same node.

Reviewer 2: ```suggestion
The xref:airflow::index.adoc[Airflow operator] can use a xref:airflow::usage.adoc#_via_persistentvolumeclaim[PersistentVolumeClaim for externally-loaded airflow jobs], known as Direct Acyclic Graphs (DAGs), as shown in this https://github.com/stackabletech/airflow-operator/blob/main/examples/simple-airflow-cluster-dags-pvc.yaml[example]. Airflow expects that all components (webserver, scheduler and workers) have access to the DAG folder, so the PersistentVolumeClaim either has to be accessible with the `ReadWriteMany` access mode or node selection should be declared to ensure all components run on the same node.
```

=== Spark-k8s

> Users of the Spark-k8s operator have a variety of ways to manage SparkApplication dependencies, one of which is to mount resources on a PersistentVolumeClaim. An example is shown https://github.com/stackabletech/spark-k8s-operator/blob/main/examples/ny-tlc-report.yaml[here].

Reviewer 2: ```suggestion
Users of the xref:spark-k8s::index.adoc[Spark-k8s operator] have a variety of ways to manage SparkApplication dependencies, one of which is to xref:spark-k8s::usage.adoc#_pyspark_externally_located_dataset_artifact_available_via_pvcvolume_mount[mount resources on a PersistentVolumeClaim]. An example is shown https://github.com/stackabletech/spark-k8s-operator/blob/main/examples/ny-tlc-report.yaml[here].
```
